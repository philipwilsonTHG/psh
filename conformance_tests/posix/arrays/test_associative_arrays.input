# Test associative array support
# PSH supports bash-style associative arrays with declare -A

echo "Testing associative arrays:"

# Test basic associative array declaration and assignment
echo "Testing basic associative operations:"
declare -A assoc_array

assoc_array["key1"]="value1"
assoc_array["key2"]="value2"
assoc_array["key3"]="value3"

echo "assoc_array[key1]: ${assoc_array[key1]}"
echo "assoc_array[key2]: ${assoc_array[key2]}"
echo "assoc_array[key3]: ${assoc_array[key3]}"

# Test associative array initialization
echo "Testing associative array initialization:"
declare -A colors=([red]="#FF0000" [green]="#00FF00" [blue]="#0000FF")

echo "colors[red]: ${colors[red]}"
echo "colors[green]: ${colors[green]}"
echo "colors[blue]: ${colors[blue]}"

# Test accessing all values and keys
echo "Testing array expansion:"
echo "All values: ${colors[@]}"
echo "All keys: ${!colors[@]}"

# Test array length
echo "Testing array length:"
echo "Length of colors: ${#colors[@]}"
echo "Length of assoc_array: ${#assoc_array[@]}"

# Test string keys with spaces and special characters
echo "Testing special keys:"
declare -A special_keys

special_keys["key with spaces"]="spaced value"
special_keys["key-with-dashes"]="dashed value"
special_keys["key_with_underscores"]="underscore value"
special_keys["key.with.dots"]="dotted value"
special_keys["key@symbol"]="symbol value"

echo "Key with spaces: ${special_keys[key with spaces]}"
echo "Key with dashes: ${special_keys[key-with-dashes]}"
echo "Key with underscores: ${special_keys[key_with_underscores]}"
echo "Key with dots: ${special_keys[key.with.dots]}"
echo "Key with symbol: ${special_keys[key@symbol]}"

# Test accessing all special keys
echo "All special keys: ${!special_keys[@]}"
echo "All special values: ${special_keys[@]}"

# Test key modification and deletion
echo "Testing key modification:"
config=()
declare -A config
config["host"]="localhost"
config["port"]="8080"
config["debug"]="true"

echo "Original config: ${config[@]}"
echo "Config keys: ${!config[@]}"

# Modify existing key
config["port"]="9090"
echo "After port change: ${config[port]}"

# Add new key
config["ssl"]="false"
echo "After adding ssl: ${config[@]}"

# Delete a key
unset config["debug"]
echo "After removing debug: ${config[@]}"
echo "Keys after removal: ${!config[@]}"

# Test complex key expressions
echo "Testing complex keys:"
declare -A complex

var1="dynamic"
var2="key"
complex["${var1}_${var2}"]="dynamic value"
complex["$(echo computed)"]="computed value"

echo "Dynamic key: ${complex[dynamic_key]}"
echo "Computed key: ${complex[computed]}"

# Test associative array in loops
echo "Testing associative array iteration:"
declare -A fruits=([apple]="red" [banana]="yellow" [grape]="purple" [orange]="orange")

echo "Iterating over values:"
for fruit_color in "${fruits[@]}"; do
    echo "  Color: $fruit_color"
done

echo "Iterating over keys:"
for fruit_name in "${!fruits[@]}"; do
    echo "  Fruit: $fruit_name"
done

echo "Iterating over key-value pairs:"
for fruit in "${!fruits[@]}"; do
    echo "  $fruit is ${fruits[$fruit]}"
done

# Test nested-like behavior with concatenated keys
echo "Testing pseudo-nested arrays:"
declare -A nested

nested["user.name"]="john"
nested["user.email"]="john@example.com"
nested["user.age"]="30"
nested["config.debug"]="true"
nested["config.verbose"]="false"

echo "User info:"
echo "  Name: ${nested[user.name]}"
echo "  Email: ${nested[user.email]}"
echo "  Age: ${nested[user.age]}"

echo "Config info:"
echo "  Debug: ${nested[config.debug]}"
echo "  Verbose: ${nested[config.verbose]}"

# Test key existence checking
echo "Testing key existence:"
declare -A test_exist=([exists]="yes" [another]="value")

if [[ -v test_exist[exists] ]]; then
    echo "Key 'exists' is set"
else
    echo "Key 'exists' is not set"
fi

if [[ -v test_exist[missing] ]]; then
    echo "Key 'missing' is set"
else
    echo "Key 'missing' is not set"
fi

# Test with empty values
echo "Testing empty values:"
declare -A with_empty
with_empty["empty_key"]=""
with_empty["null_key"]=""
with_empty["normal_key"]="normal value"

echo "Empty key: [${with_empty[empty_key]}]"
echo "Null key: [${with_empty[null_key]}]"
echo "Normal key: [${with_empty[normal_key]}]"
echo "All keys: ${!with_empty[@]}"

# Test array copying and assignment
echo "Testing array copying:"
declare -A original=([a]="1" [b]="2" [c]="3")
declare -A copy

# Copy elements manually (bash doesn't support direct associative array copying)
for key in "${!original[@]}"; do
    copy["$key"]="${original[$key]}"
done

echo "Original: ${original[@]}"
echo "Copy: ${copy[@]}"

# Modify copy to verify independence
copy["b"]="modified"
echo "After modifying copy:"
echo "Original: ${original[@]}"
echo "Copy: ${copy[@]}"

# Test associative arrays with functions
echo "Testing associative arrays in functions:"
process_assoc_array() {
    local -A local_array=([func_key1]="func_val1" [func_key2]="func_val2")
    echo "Local associative array: ${local_array[@]}"
    echo "Local keys: ${!local_array[@]}"
    
    # Access global associative array
    echo "Global array from function: ${global_assoc[@]}"
}

declare -A global_assoc=([global1]="value1" [global2]="value2")
echo "Before function call: ${global_assoc[@]}"
process_assoc_array

# Test associative array with numeric-like keys
echo "Testing numeric-like string keys:"
declare -A numeric_keys
numeric_keys["1"]="first"
numeric_keys["2"]="second"
numeric_keys["10"]="tenth"
numeric_keys["01"]="zero-one"

echo "Key '1': ${numeric_keys[1]}"
echo "Key '2': ${numeric_keys[2]}"
echo "Key '10': ${numeric_keys[10]}"
echo "Key '01': ${numeric_keys[01]}"
echo "All numeric keys: ${!numeric_keys[@]}"

# Test case sensitivity
echo "Testing case sensitivity:"
declare -A case_test
case_test["Key"]="uppercase K"
case_test["key"]="lowercase k"
case_test["KEY"]="all uppercase"

echo "Key: ${case_test[Key]}"
echo "key: ${case_test[key]}"
echo "KEY: ${case_test[KEY]}"

# Test error conditions
echo "Testing error handling:"
declare -A error_test=([good]="value")

echo "Valid key: ${error_test[good]}"
echo "Invalid key: ${error_test[nonexistent]}"

# Test array clearing
echo "Testing array clearing:"
declare -A to_clear=([clear1]="value1" [clear2]="value2")
echo "Before clearing: ${to_clear[@]}"
unset to_clear
echo "After unset: ${to_clear[@]}"

# Test mixed key types
echo "Testing mixed key types:"
declare -A mixed
mixed["string_key"]="string value"
mixed["123"]="numeric string key"
mixed[""]="empty key"

echo "String key: ${mixed[string_key]}"
echo "Numeric string key: ${mixed[123]}"
echo "Empty key: ${mixed[]}"
echo "All mixed keys: ${!mixed[@]}"

echo "Associative arrays test completed"