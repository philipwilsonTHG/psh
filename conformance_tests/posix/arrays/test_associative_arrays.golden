Testing associative arrays:
Testing basic associative operations:
assoc_array[key1]: value1
assoc_array[key2]: value2
assoc_array[key3]: value3
Testing associative array initialization:
colors[red]: #FF0000
colors[green]: #00FF00
colors[blue]: #0000FF
Testing array expansion:
All values: #FF0000 #00FF00 #0000FF
All keys: red green blue
Testing array length:
Length of colors: 3
Length of assoc_array: 3
Testing special keys:
Key with spaces: spaced value
Key with dashes: dashed value
Key with underscores: underscore value
Key with dots: dotted value
Key with symbol: symbol value
All special keys: key with spaces key-with-dashes key_with_underscores key.with.dots key@symbol
All special values: spaced value dashed value underscore value dotted value symbol value
Testing key modification:
Original config: localhost 8080 true
Config keys: host port debug
After port change: 9090
After adding ssl: localhost 9090 true false
After removing debug: localhost 9090 false
Keys after removal: host port ssl
Testing complex keys:
Dynamic key: dynamic value
Computed key: computed value
Testing associative array iteration:
Iterating over values:
  Color: red
  Color: yellow
  Color: purple
  Color: orange
Iterating over keys:
  Fruit: apple
  Fruit: banana
  Fruit: grape
  Fruit: orange
Iterating over key-value pairs:
  apple is red
  banana is yellow
  grape is purple
  orange is orange
Testing pseudo-nested arrays:
User info:
  Name: john
  Email: john@example.com
  Age: 30
Config info:
  Debug: true
  Verbose: false
Testing key existence:
Key 'exists' is set
Key 'exists' is not set
Key 'missing' is set
Key 'missing' is not set
Testing empty values:
Empty key: []
Null key: []
Normal key: [normal value]
All keys: empty_key null_key normal_key
Testing array copying:
Original: 1 2 3
Copy: 1 2 3
After modifying copy:
Original: 1 2 3
Copy: 1 modified 3
Testing associative arrays in functions:
Before function call: value1 value2
Local associative array: ([func_key1]=func_val1 [func_key2]=func_val2)
Local keys: 0
Global array from function: value1 value2
Testing numeric-like string keys:
Key '1': first
Key '2': second
Key '10': tenth
Key '01': zero-one
All numeric keys: 1 2 10 01
Testing case sensitivity:
Key: uppercase K
key: lowercase k
KEY: all uppercase
Testing error handling:
Valid key: value
Invalid key:
Testing array clearing:
Before clearing: value1 value2
After unset:
Testing mixed key types:
String key: string value
Numeric string key: numeric string key
Empty key: empty key
All mixed keys: string_key 123
Associative arrays test completed
